google-cloud-storage/lib/google/cloud/storage/file/signer_v4.rb

```diff
@@ -122,9 +122,10 @@ def issuer_and_signer issuer, client_email, signing_key, private_key
           def canonical_and_signed_headers headers
             # Headers needs to be in alpha order.
             canonical_headers = headers || {}
-            canonical_headers = Hash[canonical_headers.map do |k, v|
-                                       [k.downcase, v.strip.gsub(/\s+/, " ")]
-                                     end]
+            headers_arr = canonical_headers.map do |k, v|
+              [k.downcase, v.strip.gsub(/[^\S\t]+/, " ").gsub(/\t+/, "\t")]
```

Comment:
@jskeet Does this whitespace replacement look ok to you? It passes the conformance tests.

Reply:
Can't look now - will check tomorrow

Reply:
Hmm... I'm not sure. In the C# code we collapse whitespace runs by replacing any run of whitespace with "the final character of that run"... which seems equally somewhat arbitrary, to be honest. (So "space tab" ends up as tab, but "tab space" ends up as space.)
I don't remember seeing details about this in the spec we were following - it may be worth following up on this. @frankyn Any thoughts?

Reply:
Hi folks,
IIUC the problem breaks down to the following two questions:
Question 1: Which of the following is the correct result? Based on development the expected result is option 1.
1. `[tab][tab] -> [tab] && [space][space] -> [space]`
1. `[tab][tab] -> [space] && [space][space] -> [space]`

Question 2: Which of the following is the correct result? I'm not sure, but based on Question 1, I'd select option 2 to the expected result.
1. `[tab][tab][space][space] -> [tab][space]`
1. `[tab][tab][space][space] -> [space] && [space][space][tab][tab] -> [space]` 
1. `[tab][tab][space][space] -> [space] && [space][space][tab][tab] -> [tab]`

Background on handling whitespace wasn't well defined in the spec given the source material used: "Convert sequential spaces in the header value to a single space." is the documented expectation. 

If I'm on the same page as both of you (@jskeet, @quartzmo), PLMK and I'll follow-up with the GCS team.

Reply:
From my implementation, I'd expect option 3 for your final set - but we really need to know what the server expects, and document that.

Reply:
Thanks @jskeet. Sg, so I'm on the same page and will follow-up with the GCS team.

Reply:
Internal bug: 136171758

Reply:
@jskeet Can we add the `[tab][tab][space][space]` scenario to the conformance test?

Reply:
@quartzmo, as to not block HMAC SA support, can we move v4 signature issues into a separate PR?

Reply:
@frankyn any word on this?

Reply:
@jdpedrie, I pinged the GCS team last week but I haven't heard back yet. I would recommend not blocking support for HMAC SA for this question and splitting it off into a separate PR instead.